embassy-mspm0/src/watchdog.rs

```diff
@@ -0,0 +1,330 @@
+            | Self::MSec500
+            | Self::Sec4
+            | Self::Sec32
+            | Self::Sec256
+            | Self::Sec4096 => 3u8,
+            //  divide by 5
+            Self::USec9770
+            | Self::USec39060
+            | Self::USec156250
+            | Self::MSec630
+            | Self::Sec5
+            | Self::Sec40
+            | Self::Sec320
+            | Self::Sec5120 => 4u8,
+            //  divide by 6
+            Self::USec11720
+            | Self::USec46880
+            | Self::USec187500
+            | Self::MSec750
+            | Self::Sec6
+            | Self::Sec48
+            | Self::Sec384
+            | Self::Sec6144 => 5u8,
+            //  divide by 7
+            Self::USec13670
+            | Self::USec54690
+            | Self::USec218750
+            | Self::MSec880
+            | Self::Sec7
+            | Self::Sec56
+            | Self::Sec448
+            | Self::Sec7168 => 6u8,
+            //  divide by 8
+            Self::Sec512 | Self::Sec8192 => 7u8,
+        }
+    }
+}
+
+/// Timeout percentage that is treated as "too early" and generates violation.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum ClosedWindowPercentage {
+    // window period is not used
+    Zero,
+    // 12.5% percents
+    Twelve,
+    // 18.75% percents
+    Eighteen,
+    // 25% percents
+    TwentyFive,
+    // 50% percents
+    Fifty,
+    // 75% percents
+    SeventyFive,
+    // 81.25% percents
+    EightyOne,
+    // 87.5% percents
+    EightySeven,
+}
+
+#[non_exhaustive]
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+/// Watchdog Config
+pub struct Config {
+    /// Watchdog timeout
+    pub timeout: Timeout,
+
+    /// closed window percentage
+    pub closed_window: ClosedWindowPercentage,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            timeout: Timeout::Sec1,
+            closed_window: ClosedWindowPercentage::Zero,
+        }
+    }
+}
+
+pub struct Watchdog<'d, T: Instance> {
+    wdt: PhantomData<&'d mut T>,
+}
+
+impl<'d, T: Instance> Watchdog<'d, T> {
+    /// Watchdog initialization.
+    pub fn new(_instance: Peri<'d, T>, config: Config) -> Self {
+        // Init power for watchdog
+        T::REGS.gprcm(0).rstctl().write(|w| {
+            w.set_resetstkyclr(true);
+            w.set_resetassert(true);
+            w.set_key(vals::ResetKey::KEY);
+        });
+
+        // Enable power for watchdog
+        T::REGS.gprcm(0).pwren().write(|w| {
+            w.set_enable(true);
+            w.set_key(vals::PwrenKey::KEY);
+        });
+
+        // init delay, 16 cycles
+        cortex_m::asm::delay(16);
+
+        //init watchdog
+        T::REGS.wwdtctl0().write(|w| {
+            w.set_clkdiv(config.timeout.get_clkdiv());
+            w.set_per(config.timeout.get_period());
+            w.set_mode(vals::Mode::WINDOW);
+            match config.closed_window {
```

Comment:
I think a conversion function for closed_window would make the function a little less long.